libia2/threads.c

```diff
@@ -18,8 +18,8 @@ void *ia2_thread_begin(void *arg) {
   /* Free the thunk. */
   munmap(arg, sizeof(struct ia2_thread_thunk));
 
-  protect_tls();
   init_stacks();
+  late_protect_tls();
```

Comment:
We should document the ordering requirements here.

Specifically, `late_protect_tls` comes after `init_stacks` because it wants to switch to those stacks as part of call-gates.

However, we have the problem that the previous ordering was also necessary: `init_stacks` writes initial stack pointers to the per-compartment `ia2_stackptr_N` TLS variables. If these haven't been `pkey_mprotected`, then any rogue compartment can overwrite them between when we set up stacks and when we actually try to use them.

`init_stacks` and the `ALLOCATE_COMPARTMENT_STACK` macro that it uses are specifically written to not require stacks to be set up yet when they run, and the macro changes pkeys and allocate per-compartment stacks without needing a call-stack.

Looking at `late_protect_tls`, it simply calls the right set of `late_init_tls_N()` functions, which does `dl_iterate_phdr(protect_tls_pages, &args);`. `protect_tls_pages` does `dlsym`, some logic, and then `pkey_mprotect`. The only part of this that should actually be compartment-dependent is `dlsym`, and that's what the backtrace shows. But what I understand here is that we need to call `dlsym` from the compartment for libdl, not the compartment whose TLS pages we're protecting?

Is that what we're doing by wrapping all the `late_init_tls_N` functions in call-gates? Why not just wrap `late_protect_tls`?

Reply:
I talked about this with @rinon. Basically, [`late_`]`protect_tls` needs to run each `late_protect_tls_for_compartment` with the appropriate pkey so that `dl_iterate_phdr` can access the stack. That stack switching has to happen before we set up `ia2_stackptr_N` (because they live in TLS), so we need to allocate and switch to a stack for each compartment in order to be able to run `late_protect_tls_for_compartment` for it. This sounds like a dependency loop, but it isn't because we don't actually need `ia2_stackptr_N` to be set up until we real call-gate, and this sequence can just be hard-coded control flow without call gates. We can even save the stack we used to call `late_protect_tls_for_compartment` in `ia2_stackptr_N` after `late_protect_tls_for_compartment` ends.

I'll see if I can implement this atop this PR.

Reply:
> But what I understand here is that we need to call dlsym from the compartment for libdl

Did we decide we need to put libdl in its own compartment? Would this be optional or required?

> I'll see if I can implement this atop this PR.
@fw-immunant if you're making changes on top of this, feel free to push to this branch/PR if it'd make review easier

Reply:
I don't think we necessarily have to put libdl in its own compartment. I think we can entirely avoid libdl in `protect_tls_pages` if we either pass in the address of `ia2_stackptr_0` or hard-code `&ia2_stackptr_0` as the page to avoid in protect_tls_pages. I've done the former but I'm not completely sure about the security of that approach--is there a way to prevent this being a useful gadget? See `PhdrSearchArgsTls`--cc @rinon

Reply:
Can't you just use `ia2_stackptr_0` as a PC-relative reference inside of `protect_tls_pages`? That should be safe, as you can only ever refer to that label, not an arbitrary address.

Reply:
We can do that. I was curious whether the approach we were taking in #248 would also help us with TLS setup (and allow us to not hard-code `ia2_stackptr_0` in `protect_tls_pages`, possibly helping with shared TLS later on) but I'm not sure that it would. I also think we can probably also eliminate the `pkey` field from `struct PhdrSearchArgs` as of that PR.